Custom status expiry

manojmalik20 opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A brief description of what this pull request does.
-->

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket or fixes a reported issue, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-mobile/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

#### Checklist
<!--
Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.
-->
- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates

#### Device Information
This PR was tested on: <!-- Device name(s), OS version(s) -->

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs/Videos (for both iOS and Android if possible).
-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* New features and improvements, including behavioural changes, UI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Example:

```release-note
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->

```release-note

```

manojmalik20 committed:
Started with the custom status feature
Added custom status actions and API integration
Added custom status selectors and types
Added custom status emoji component
Added EnableCustomStatuses flag in config
Added Set a status option in the settings sidebar

manojmalik20 committed:
Fixed emoji and added clear button
Fixed the default emoji in the set a status in the settings sidebar
Made a custom status label component to match the styling of rest of the drawer items
Fixed the bug in the selectors
Added localization id in en.json

manojmalik20 committed:
Added icon for clear button
Fixed the custom status label issue
Added styling and icon for the clear button
Fixed order of the localization ids

manojmalik20 committed:
Made custom status modal with suggestions
Fixed the long custom status overflowing in the sidebar with ellipses
Added the functionality to open custom status modal from sidebar
Made the constants for custom status
Made the custom status suggestion functional component
Made the custom status modal functional component
Made the selector for the recent custom statuses
Added some localization ids

manojmalik20 committed:
Changed the custom status modal to class component
Added recent custom statuses in the modal
Fixed the styling of titles
Fixed lint errors except one

chetanyakan committed:
Fix types and use i18n translations

chetanyakan committed:
Update types

manojmalik20 committed:
Refactored some code and made clear button component
Refactored the code of the custom status label to a different file and used it everywhere
Made a separate component for clear button and added it in recents and sidebar
Fixed the styling of the custom status text according to the new figma designs

manojmalik20 committed:
Added emojipicker in custom status modal
Fixed the Done button in custom status modal
Added the functionality to clear status from inside the modal

manojmalik20 committed:
Merge pull request #2 from brightscout-alpha/MI-1190

Mi 1190

manojmalik20 committed:
Added custom status in user profile and post header and review fixes
Added custom status in user profile page
Added custom status emoji in post header
Added functionality to preventDoubleTap everywhere
Added localization ids
Review fixes

manojmalik20 committed:
Fixed several issues
Increased touch area of clear button everywhere
Fixed the event propagation issue in clearing suggestions
Fixed the center alignment issue of title in custom status modal
Added clear button in user profile page
Review fixes

chetanyakan committed:
Fix types for custom status modal

manojmalik20 committed:
Merge branch 'MI-1190' of github.com:brightscout-alpha/mattermost-mobile into MI-1190

chetanyakan committed:
Fix tsc

manojmalik20 committed:
Fixed types and separated functions for Recent and suggestions in custom status modal

chetanyakan committed:
Merge branch 'MI-1190' of github.com:brightscout-alpha/mattermost-mobile into MI-1190

* 'MI-1190' of github.com:brightscout-alpha/mattermost-mobile:
  Fixed types and separated functions for Recent and suggestions in custom status modal

chetanyakan committed:
Update app/components/sidebars/settings/settings_sidebar_base.js

manojmalik20 committed:
Updated localization id in settings sidebar

manojmalik20 committed:
Merge pull request #1 from brightscout-alpha/MI-1167

Started with the custom status feature

manojmalik20 committed:
Changed the method in removeRecentCustomStatus API call
Changed names of some variables

manojmalik20 committed:
Merge branch 'master' of github.com:brightscout-alpha/mattermost-mobile into MI-1190

manojmalik20 committed:
Fixed the input height issue in ios and modalContext issue in ios

manojmalik20 committed:
Fixed various issues
Fixed clear button border radius issue in ios
Changed the method to post in removeRecentCustomStatus API call
Made the custom status modal input single line and fixed the height
Changed the height of iconContainer in custom status suggestion

chetanyakan committed:
Merge branch 'custom-status' of github.com:brightscout-alpha/mattermost-mobile into custom-status

* 'custom-status' of github.com:brightscout-alpha/mattermost-mobile:
  Added emojipicker in custom status modal Fixed the Done button in custom status modal Added the functionality to clear status from inside the modal
  Refactored some code and made clear button component Refactored the code of the custom status label to a different file and used it everywhere Made a separate component for clear button and added it in recents and sidebar Fixed the styling of the custom status text according to the new figma designs
  Update types
  Fix types and use i18n translations
  Changed the custom status modal to class component Added recent custom statuses in the modal Fixed the styling of titles Fixed lint errors except one
  Made custom status modal with suggestions Fixed the long custom status overflowing in the sidebar with ellipses Added the functionality to open custom status modal from sidebar Made the constants for custom status Made the custom status suggestion functional component Made the custom status modal functional component Made the selector for the recent custom statuses Added some localization ids
  Added icon for clear button Fixed the custom status label issue Added styling and icon for the clear button Fixed order of the localization ids
  Fixed emoji and added clear button Fixed the default emoji in the set a status in the settings sidebar Made a custom status label component to match the styling of rest of the drawer items Fixed the bug in the selectors Added localization id in en.json
  Started with the custom status feature Added custom status actions and API integration Added custom status selectors and types Added custom status emoji component Added EnableCustomStatuses flag in config Added Set a status option in the settings sidebar

chetanyakan committed:
Merge branch 'custom-status' into MI-1190

chetanyakan committed:
Merge pull request #3 from brightscout-alpha/MI-1190

- Added custom status in user profile page
- Added custom status emoji in post header
- Added functionality to preventDoubleTap everywhere
- Added localization ids

manojmalik20 committed:
Fixed the event propagation issue without use of PanResponder

manojmalik20 committed:
Unit tests

Added unit tests for custom status emoji component
Added unit tests for custom status text component
Added unit tests for clear button component
Added unit tests for custom status suggestion component
Added unit tests for custom status modal component
Added tests and updated snapshtots in user profile screen

manojmalik20 committed:
Added custom status in several new locations
Added custom status emoji in LHS channel list
Added custom status emoji in channel header
Added custom status in channel info header
Modified the styles of some components

manojmalik20 committed:
Merge pull request #7 from brightscout-alpha/MI-1200

Added custom status in several new locations

manojmalik20 committed:
Fixed width issue in LHS channel list and channel title (#8)

Added one check in channel info header

manojmalik20 committed:
E2E tests for custom status (#6)

* Completed MM-T3890, MM-T3891, MM-T3892, MM-T3893 e2e tests

Added EnableCustomUserStatuses in default config of detox
Added several testIDs in custom status modal and clear button components
Modified settings_sidebar ui component in detox support
Made custom status screen ui screen in detox support
Added testIDs in custom status suggestion and clear buttons
Modified custom status ui screen component in detox support
Added appropriate comments in the tests
Modified custom status emoji component to accept styling and added testID
Modified the custom status emoji styling in post header, user profile, LHS channel list, channel info header, channel title
Made user profile screen in detox support
Added testID prop in custom status emoji component
Added testIDs in channel info header and user profile and user profile row components

manojmalik20 committed:
Added unit tests for new components and updated tests for custom status emoji

Updated snapshots for existing components
Added new tests for channel title and LHS channel item
Added new tests for channel info and channel info header
Updated snapshots

manojmalik20 committed:
Fixed styles and emoji opacity in custom status input in modal

* Fixed the color of text and emoji in custom status suggestions
* Fixed the color of emoji and font size in the custom status input in modal

manojmalik20 committed:
Updated snapshots

manojmalik20 committed:
Fix the MM-T3893 E2E test case

* Update detox/e2e/test/smoke_test/custom_status.e2e.js

chetanyakan committed:
Apply suggestions from code review

Co-authored-by: Joseph Baylon <joseph.baylon@mattermost.com>

manojmalik20 committed:
Merge branch 'master' of github.com:brightscout-alpha/mattermost-mobile into custom-status

manojmalik20 committed:
Updated snapshots and moved the custom status e2e test to different location

availchet committed:
Fix styles

availchet committed:
Fix ios styles

availchet committed:
Update snapshots

manojmalik20 committed:
Updated styles and localization ids

manojmalik20 committed:
Changed the API call for the remove recent custom status action (#12)

availchet committed:
Fix styles for ios

availchet committed:
Merge branch 'custom-status' of github.com:brightscout-alpha/mattermost-mobile into custom-status

* 'custom-status' of github.com:brightscout-alpha/mattermost-mobile:
  Changed the API call for the remove recent custom status action (#12)
  Updated styles and localization ids

manojmalik20 committed:
Fixed styles in Android
Fixed the styling of clear button everywhere
Fixed the separator not showing issue in custom status suggestion
Fixed the emoji positioning issue in custom status suggestion

manojmalik20 committed:
Merge branch 'master' of github.com:brightscout-alpha/mattermost-mobile into custom-status

manojmalik20 committed:
Fixed tests and updated snapshots

availchet committed:
Fix stylessss for ios

manojmalik20 committed:
Updated snapshots

manojmalik20 committed:
Added close button and title in emojipicker screen opened from custom status modal

manojmalik20 committed:
Merge branch 'master' of https://github.com/mattermost/mattermost-mobile into custom-status

manojmalik20 committed:
Fixed the delay issue in clearing and setting the custom status in settings sidebar (#13)

* Fixed the delay issue in clearing the custom status in settings sidebar
Modified the unsetCustomStatus action to use request statuses
Modified the settings sidebar to use the request status to show or hide custom status
Made proper types and reducers for the unset custom status request

* Added the Retry message for setting and clearing the custom status
Made set custom status request action types and reducers
Modified the setCustomStatus action to dispatch request status actions
Made the retryMessage component
Added the logic for when to show the retry message

* Modified set custom status action to dispatch Update me action with the new custom status

* Modified the settings sidebar to use componentDidUpdate for the delay issue in custom status clearing and setting

* Modified componentDidUpdate and made a new function for handling request status change

* Modified handleRequestStatusChange function to two separate functions

* Code refactoring

manojmalik20 committed:
Updated tests

manojmalik20 committed:
Merge branch 'master' of github.com:brightscout-alpha/mattermost-mobile into custom-status

manojmalik20 committed:
Fixed the issue for custom status not updating in settings sidebar when changed from another client

* Fixed the issue for custom status not updating in settings sidebar when changed from another client
* Fixed the IOS UI issues
    - Fixed the font-weight of the channel header custom status
    - Fixed the extra space added on typing in input issue
* Fix custom status e2e tests

availchet committed:
Merge branch 'master' of github.com:mattermost/mattermost-mobile into custom-status

* 'master' of github.com:mattermost/mattermost-mobile:
  Bump app build number to 353 (#5331)
  MM-30429 Detox/E2E: Added e2e for MM-T3190, MM-T3212, MM-T3214, MM-T3227 (#5276)
  MM-27849 render emojis with appended skin tone (#5297)
  Fix at_mention and channel_mention autocomplete on iOS (#5327)
  Refactor rest client class to use mixin and a bit of cleanup (#5318)
  Bump app build number to 352 (#5324)
  Set url trailing slash (#5322)
  Use RNN 7.11.3 (#5317)
  MM-30431 Detox/E2E: Added e2e for MM-T3196, MM-T3210, MM-T3211, MM-T3262, MM-T3264 (#5313)
  Bump app build number to  351 (#5309)
  Translations update from Weblate (#5307)

manojmalik20 committed:
Fixed the custom status input click area issue

availchet committed:
Fix custom status input touchable area

manojmalik20 committed:
Merge branch 'master' of github.com:brightscout-alpha/mattermost-mobile into custom-status

manojmalik20 committed:
added expiry time to custom status

manojmalik20 committed:
removed spaces from en.json

manojmalik20 committed:
Merge branch 'master' of github.com:mattermost/mattermost-mobile into custom-status

manojmalik20 committed:
Updated snapshots

manojmalik20 committed:
Merge branch 'master' of github.com:mattermost/mattermost-mobile into custom-status

manojmalik20 committed:
Updated snapshots

manojmalik20 committed:
added some ui changes to expiry stuff in custom status

manojmalik20 committed:
Apply suggestions from code review

Co-authored-by: Michel Engelen <32863416+michelengelen@users.noreply.github.com>

manojmalik20 committed:
Review fixes (#15)

* Review fixes
Sorted the import order in all files
Modified removeRecentCustomStatus action not to use bindClientFunction
Added types for custom status client functions
Added containerSize prop in clear button component
Removed the check for custom status enabled from custom status emoji and added it in post_header, channel_title, channel_item
Modified getCustomStatusSelector to return undefined if custom status does not exist
Modified settings_sidebar, channel_info, channel_info_header, custom_status_modal, user_profile to handle undefined custom status
Removed the text wrapper from the custom status emoji in channel_info_header
Added check for identifying the navigation button pressed in custom status modal

* Added server version check in isCustomStatusEnabled selector

* Refactored some code and added types in style in custom status modal
Added type in custom status suggestion functions in custom status modal
Refactored some code

manojmalik20 committed:
Fixed failing tests and updated snapshots

manojmalik20 committed:
Expiry Time functionality added

manojmalik20 committed:
Updated snapshot for clear button component

manojmalik20 committed:
Merge branch 'custom-status' of code.jtg.tools:brightscout/mattermost-mobile into test-1230

manojmalik20 committed:
ran npm run fix

manojmalik20 committed:
Added timezone functionality to custom status expiry

manojmalik20 committed:
Merged custom status into test-1230

manojmalik20 committed:
Merge branch 'custom-status' of code.jtg.tools:brightscout/mattermost-mobile into test-1230

manojmalik20 committed:
Review fixes

Removed logic for tracking requests for setting and unsetting custom status
Added logic for clearing status without delay in settings sidebar
Added event for set custom status failure and attached listener in settings sidebar
Added feature in custom status text component to accept string or FormattedText as prop and ellipsizeMode and numberOfLines as well
Removed labelSibling and failureText props from drawer item
Passed clearButton and retryMessage in labelComponent from settings sidebar to drawer item
Replaced Text with CustomStatusText in channel info header
Added localization in the custom status suggestions
Added exception handling in the getRecentCustomStatuses selector

manojmalik20 committed:
Fixed drawer item unit tests and updated snapshots

manojmalik20 committed:
Merge branch 'custom-status' of https://github.com/brightscout-alpha/mattermost-mobile into custom-status-expiry

manojmalik20 committed:
Fixed UI issues for IOS in settings sidebar and main sidebar

manojmalik20 committed:
Updated snapshots

hrishavkmr commented on app/components/custom_status/custom_status_emoji.tsx:
```diff
@@ -29,15 +28,17 @@ const CustomStatusEmoji = ({emojiSize, userID, style, testID}: ComponentProps) =
+                    size={emojiSize}
+                    emojiName={customStatus.emoji}
+                />
+            </Text>
```
This render logic should be outside the `return()` in a variable.
```code
const customEmoji = customStatus ? <Text>...</Text> : null;
return (
     {customEmoji}
)
```

hrishavkmr commented on app/components/custom_status/custom_status_expiry.tsx:
```diff
@@ -0,0 +1,136 @@
+        />
+    );
+
+    const showDate = !isToday && !useTime && !useDay && (
```
condition can be simplified
```suggestion
    const showDate = !(isToday || useTime || useDay) && (
```

hrishavkmr commented on app/screens/custom_status_clear_after/date_time_selector.tsx:
```diff
@@ -0,0 +1,103 @@
+                    color={theme.buttonBg}
+                />
+            </View>
+            {show && (
+                <DateTimePicker
+                    testID='dateTimePicker'
+                    value={moment(date).toDate()}
+                    mode={mode}
+                    is24Hour={militaryTime}
+                    display={Platform.OS === 'ios' ? 'spinner' : 'default'}
+                    onChange={onChange}
+                />
+            )}
```
store this outside the render function in a variable and render that variable here.

hrishavkmr commented on app/screens/custom_status/custom_status_suggestion.tsx:
```diff
@@ -56,11 +61,22 @@ const CustomStatusSuggestion = (props: Props) => {
+                                <CustomStatusText
+                                    text={intl.formatMessage(durationValues[duration])}
+                                    theme={theme}
+                                    textStyle={{color: changeOpacity(theme.centerChannelColor, 0.6)}}
```
create a new style in the stylesheet and use that style here.

manojmalik20 committed:
Refactored duration code in custom status suggestion

manojmalik20 committed:
Review fixes
Removed check for user null or undefined from setCustomStatus action
Refactored some code in custom status emoji and post header
Memoized handleClear in custom status suggestion
Made stylesheet in settings_sidebar and added all styles there
Consolidated openCustomStatusModal function inside goToCustomStatusScreen

manojmalik20 committed:
Merge branch 'master' of github.com:mattermost/mattermost-mobile into custom-status

manojmalik20 committed:
Refactored some code
Refactored some code in set custom status action
Updated name of a unit test in drawer item and updated snapshot

manojmalik20 committed:
added ui fixes to date time picker component and displayExpiryTime
component

manojmalik20 committed:
Removed server version check for testing

manojmalik20 committed:
Merge branch 'custom-status' of code.jtg.tools:brightscout/mattermost-mobile into custom-status-expiry

manojmalik20 committed:
Merge branch 'master' of github.com:mattermost/mattermost-mobile into custom-status

manojmalik20 committed:
bug fix with timezone and statusSame

manojmalik20 committed:
Review UI fixes (#18)

Changed the padding for custom status both in settings sidebar and custom status suggestions
Changed the position of clear button in the settings sidebar

manojmalik20 committed:
Merge branch 'master' of github.com:mattermost/mattermost-mobile into custom-status

manojmalik20 committed:
UI fix in user profile screen
Fixed the call to apiGetChannelByName in custom status e2e test

manojmalik20 committed:
Added server version check for custom status feature

manojmalik20 committed:
Updated snapshots

manojmalik20 committed:
modified testIDs, output rendering logic

manojmalik20 committed:
Merge branch 'custom-status-expiry' of code.jtg.tools:brightscout/mattermost-mobile into custom-status-expiry

manojmalik20 committed:
Merge branch 'custom-status' of github.com:brightscout-alpha/mattermost-mobile into custom-status-expiry

manojmalik20 committed:
Review fixes
Converted mapStateToProps to makeMapStateToProps and moved makeGetCustomStatus call inside them in several components

manojmalik20 committed:
Merge branch 'custom-status' of github.com:brightscout-alpha/mattermost-mobile into custom-status-expiry

manojmalik20 committed:
Removed the server version check and fixed showing custom status in user profile bug

manojmalik20 committed:
Merge branch 'master' of github.com:mattermost/mattermost-mobile into custom-status

manojmalik20 committed:
Updated datetime picker version and fixed timezone issue
Fixed the memoization issue in getCurrentUserTimezone selector
Refactored the getCurrentDateTimeForTimezone to getCurrentMomentForTimezone
Added the timezoneOffsetInMinutes prop to the datetime picker

manojmalik20 committed:
Fixed ios UI issues
Fixed the positioning of Clear after in custom status modal in ios
Fixed the positioning of spinner date time picker in ios

manojmalik20 committed:
Merge branch 'custom-status' of github.com:brightscout-alpha/mattermost-mobile into custom-status-expiry

manojmalik20 committed:
expiry time now showing in custom status,clear after modals

manojmalik20 committed:
Fixed the custom status not showing in user profile of other users if own status is not set

manojmalik20 committed:
Changed the emoji size in the channel title

manojmalik20 committed:
Updated snapshots

manojmalik20 committed:
Merge branch 'master' of github.com:mattermost/mattermost-mobile into custom-status

manojmalik20 committed:
Added custom status emoji in post header
Fixed the lint errors and type check errors

manojmalik20 committed:
Updated snapshots
Fixed the prop type isMilitaryTime for user_profile screen

manojmalik20 committed:
Merge branch 'custom-status' of github.com:brightscout-alpha/mattermost-mobile into custom-status-expiry

manojmalik20 committed:
UI fixes and code refactoring (#19)

* added date time picker, fixed spacing in some components

* UI fixes, added isCustomStatusExpired selector

* removed extra spaces from package-lock.json

* Review Fixes, upgraded datetimepicker version

* Refactored some code

* Refactored some code

manojmalik20 committed:
Added server version check for the custom status feature

chetanyakan commented on app/screens/channel_info/index.js:
```diff
@@ -45,13 +49,15 @@ function makeMapStateToProps() {
             }
             customStatusEnabled = isCustomStatusEnabled(state);
             customStatus = customStatusEnabled ? getCustomStatus(state, teammateId) : undefined;
```
```suggestion
            customStatus = customStatusEnabled && getCustomStatus(state, teammateId);
```

chetanyakan commented on app/components/custom_status/custom_status_expiry.tsx:
```diff
@@ -0,0 +1,139 @@
+type Props = {
+    theme: Theme;
+    time: Date;
+    styleProp?: TextStyle;
```
Rename this prop to `textStyles`

chetanyakan commented on app/components/custom_status/custom_status_expiry.tsx:
```diff
@@ -0,0 +1,139 @@
+    return (
+        <Text
+            testID={props.testID}
+            style={styleProp || styles.text}
```
I wonder if we should use something like this:

```suggestion
            style={{...styles.text, ...textStyles}}
```

chetanyakan commented on app/components/custom_status/custom_status_expiry.tsx:
```diff
@@ -0,0 +1,139 @@
+            testID={props.testID}
+            style={styleProp || styles.text}
+        >
+            {withinBrackets ? '(' : null}
```
nit -  I wonder if using `&&` here makes more sense.

```suggestion
            {withinBrackets && '('}
```

chetanyakan commented on app/screens/channel_info/channel_info_header.js:
```diff
@@ -157,6 +161,18 @@ export default class ChannelInfoHeader extends React.PureComponent {
             android: style.detail,
         });
 
+        const customStatusExpiry = customStatus?.emoji && customStatus?.duration !== CustomStatusDuration.DONT_CLEAR ?
+            (
+                <Text style={style.customStatusExpiry}>
+                    <CustomStatusExpiry
+                        time={customStatus.expires_at}
+                        theme={theme}
+                        styleProp={style.customStatusExpiry}
+                        showPrefix={true}
+                    />
+                </Text>
+            ) : null;
```
Should we add a check here for `!isCustomStatusExpired`?

manojmalik20 committed:
Refactored some code, and fixed DONT_CLEAR

manojmalik20 committed:
Merge branch 'master' of github.com:mattermost/mattermost-mobile into custom-status-expiry

manojmalik20 committed:
Merge master into custom-status-expiry

manojmalik20 committed:
Review fixes and corrected import order

manojmalik20 committed:
Resolved tsc errors, corrected intl usage

manojmalik20 committed:
Code refactoring and minor bug fixes
Fixed the default value of isCustomStatusExpired in mapStateToProps from false to true in all components
Changed the ids of various expiry dropdown texts
Fixed the order of ids in en.json

manojmalik20 committed:
Merge branch 'master' of github.com:mattermost/mattermost-mobile into custom-status-expiry

manojmalik20 committed:
Changed date time format in front of Custom option in Clear after modal (#20)

* Corrected 'check' usage in front of 'Custom' in clear-after-modal

* Renamed clear_after_suggestions to clear_after_menu_item. removed unused
code

manojmalik20 commented:
Corresponding PR has been made in mattermost.
